Introduce s3 and gdrive config

jimirocks opened the pull request.

jimirocks committed:
Introduce s3 config

jimirocks committed:
Introduce gdrive config

PetrZufan commented on manifests/config/gdrive.pp:
```diff
@@ -0,0 +1,83 @@
+      }
+        .filter |$key, $val| { $val != undef }.map |$key, $val| { "${key} ${val}" }.join(' ')
+
+      exec { default: *=> $_rclone_exec_defaults; "rclone create remote ${name} for user ${os_user}":
```
Probably 
```
... "rclone create remote ${config_name} for user ${os_user}":
```

jimirocks replied:
fixed

PetrZufan commented on manifests/config/gdrive.pp:
```diff
@@ -0,0 +1,83 @@
+        command => @("CMD")
+          rclone config create ${config_name} drive \
+          client_id ${client_id} client_secret ${client_secret} service_account_credentials ${service_account_credentials} scope ${scope} \
+          ${_options}
```
How did you resolve what is option and what is not? I mean, you can make `$_options` as
```
$_options = {
        client_id => ${client_id},
        client_secret => ${client_secret},
        service_account_credentials => ${service_account_credentials},
        scope => ${scope},
        root_folder_id => $root_folder_id,
        team_drive => $team_drive,
      }
```
and `command` 
```
command => rclone config create ${config_name} drive ${_options}
```
and it will work the same way. If I'm right.

---

Then when I look at s3.pp it's quite a same code. So maybe you can create some private resource 
```
define rclone::config {
    Enum['drive', 's3'] $backup_location,
    Hash $options,
    String $os_user,
    String $config_name,
}
```
for the common code and finaly gdrive.pp and s3.pp for setting params for this resource. 
This change will make easier implemantation of new backup locations. Let's have a call about this.

jimirocks replied:
Refactoring commit added

PetrZufan commented on manifests/config/s3.pp:
```diff
@@ -0,0 +1,97 @@
+  String $region,
+  Enum['present', 'absent'] $ensure = 'present',
+  String $config_name = $title,
+  Enum['AWS'] $s3_provider = 'AWS',
```
Is there a chance of different provider? If not, it can be private constant.

jimirocks replied:
more providers are possible, however for now not supported

PetrZufan commented on manifests/config/s3.pp:
```diff
@@ -0,0 +1,97 @@
+      }
+        .filter |$key, $val| { $val != undef }.map |$key, $val| { "${key} ${val}" }.join(' ')
+
+      exec { default: *=> $_rclone_exec_defaults; "rclone create remote ${name} for user ${os_user}":
```
Probably
```
... "rclone create remote ${config_name} for user ${os_user}":
```

jimirocks replied:
fixed

jimirocks committed:
Introduce general config

Also refactor the gdrive and s3 to use it.
Additionally making the command robust to shell-breaking sequences.
